cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -170,22 +215,65 @@ private void ensureTopicsCreated(Map<String, ?> config) {
       KafkaCruiseControlUtils.closeZkUtilsWithTimeout(zkUtils, 10000);
     }
   }
 
-  private void ensureTopicCreated(ZkUtils zkUtils, Set<String> allTopics, String topic, long retentionMs, int replicationFactor) {
+  private void increaseTopicReplicaFactor(ZkUtils zkUtils,
+                                          MetadataResponse.TopicMetadata topicMetadata,
+                                          int replicationFactor,
+                                          String topic,
+                                          Properties props) {
+    List<Integer> brokers = new ArrayList<>();
+    JavaConversions.seqAsJavaList(AdminUtils.getBrokerMetadatas(zkUtils, RackAwareMode.Safe$.MODULE$, Option.empty()))
+                   .stream().forEach(bm -> brokers.add(bm.id()));
+    if (replicationFactor > brokers.size()) {
+      throw new RuntimeException("Unable to increase topic " + topic + " replica factor to " + replicationFactor +
+                                 " since there are only " + brokers.size() + " brokers in the cluster.");
+    }
+    scala.collection.mutable.Map<Object, Seq<Object>> newReplicaAssignment = new scala.collection.mutable.HashMap<>();
+    int cursor = 0;
+    for (MetadataResponse.PartitionMetadata pm : topicMetadata.partitionMetadata()) {
+      List<Object> newAssignedReplica = new ArrayList<>();
+      // Make sure the current replicas are in new replica list.
+      pm.replicas().forEach(node -> newAssignedReplica.add(node.id()));
+      // Add new replica to partition in round-robin way.
+      while (newAssignedReplica.size() < replicationFactor) {
+        if (!newAssignedReplica.contains(brokers.get(cursor))) {
+          newAssignedReplica.add(brokers.get(cursor));
+        }
+        cursor = (cursor + 1) % brokers.size();
+      }
+      newReplicaAssignment.put(pm.partition(), JavaConverters.asScalaIteratorConverter(newAssignedReplica.iterator()).asScala().toSeq());
+    }
+    AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, newReplicaAssignment, props, true);
+  }
+
+  private void ensureTopicCreated(ZkUtils zkUtils, Set<String> allTopics, String topic, long retentionMs, int replicationFactor, int partitionCount) {
     Properties props = new Properties();
     props.setProperty(LogConfig.RetentionMsProp(), Long.toString(retentionMs));
     props.setProperty(LogConfig.CleanupPolicyProp(), DEFAULT_CLEANUP_POLICY);
     if (!allTopics.contains(topic)) {
-      AdminUtils.createTopic(zkUtils, topic, 32, replicationFactor, props, RackAwareMode.Safe$.MODULE$);
+      AdminUtils.createTopic(zkUtils, topic, partitionCount, replicationFactor, props, RackAwareMode.Safe$.MODULE$);
     } else {
       AdminUtils.changeTopicConfig(zkUtils, topic, props);
+      MetadataResponse.TopicMetadata topicMetadata = AdminUtils.fetchTopicMetadataFromZk(topic, zkUtils);
+      OptionalInt currentReplicaFactor = topicMetadata.partitionMetadata().stream().mapToInt(pm -> pm.replicas().size()).max();
```

Comment:
Two points here:
(1) I think we should get the `min()`, not the `max()` here. For example,  if I would like to set the replication factor to 3 and if this topic has 2 partitions -- one with RF 2 one with RF 3, we should increase the replication factor to 3 for the partition which has an RF 2. Currently we are ignoring such partitions.

(2) Also I suspect that we need to ensure that there is not an ongoing replica move for any partition of this topic. Because if there is a replica move, then we can see that the number of replicas for the partition is more than its actual replication factor.

Reply:
Good catch. I think replica number inconsistency for partitions of the same topic happens during replica movement due to temporal expand AR/ISR to include new replica, in this scenario the `min` replica number is the real replication factor.  so we should first ensure there is not ongoing operation then check replication factor.